### PR TITLE
IGNITE-23083 Sql. Avoid planner cache pollution with simple insert queries

### PR DESCRIPTION
**Pull request by zstan**
```
with optimization
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt     Score    Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  1011.028 ± 88.904  us/op no boost
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt     Score    Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  1024.413 ± 91.924  us/op no boost
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score    Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  313.942 ± 56.033  us/op with boost

without
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt     Score     Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  1042.638 ± 151.594  us/op no boost
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt     Score     Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  1086.291 ± 196.949  us/op no boost
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score     Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                32  avgt   20  362.045 ± 199.638  us/op with boost

with optimization
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlPreparedInsert              1    false                32  avgt   20  117.460 ± 9.242  us/op

without 
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score    Error  Units
InsertBenchmark.sqlPreparedInsert              1    false                32  avgt   20  125.572 ± 35.950  us/op
```

### CODE REVIEW

**Comment by lowka** [comment]
I think this modification should not be more complex than

```
In prepareDml:

 return validFut.thenCompose(stmt -> {
            IgniteRel optInsert = tryOptimizeInsert(stmt.value.sqlNode(), ctx.planner());
            // Check key value Modify plan that its expression list only contains literals
            // and if that is the case return the `optInsert `.
            if (optInsert != null) {
                return optInsert;
            }

           ....

```

There is no need for cache invalidation.

**Comment by korlov42** [comment]
Looks like performance on insert became worse (I double-checked these results):

```
Before
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert               1    false                25               1  avgt   20  130.512 ± 2.422  us/op

After
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                25               1  avgt   20  148.395 ± 2.009  us/op
```

**Comment by zstan** [comment]
```
opt
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score    Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                 1               1  avgt   30  303.808 ± 40.896  us/op

no opt
Benchmark                         (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score    Error  Units
InsertBenchmark.sqlInlinedInsert              1    false                 1               1  avgt   30  286.148 ± 15.264  us/op
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java [hunk 1]
```diff
@@ -678,7 +747,9 @@ private IgniteRel doOptimize(PlanningContext ctx, SqlNode validatedNode, IgniteP
             // Otherwise the cache will keep a plan that can not be used anymore,
             // and we should allow another planning attempt with increased timeout.
             if (ctx.timeouted()) {
-                cache.invalidate(key);
+                if (action != null) {
+                    action.run();
+                }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:752]: Can we just return null instead?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java [hunk 2]
```diff
@@ -197,7 +197,7 @@ static ExecutionTarget colocate(OneOfTarget oneOf, SomeOfTarget someOf) throws C
 
     static ExecutionTarget colocate(PartitionedTarget partitioned, PartitionedTarget otherPartitioned) throws ColocationMappingException {
         if (partitioned.partitionsNodes.length != otherPartitioned.partitionsNodes.length) {
-            throw new ColocationMappingException("Partitioned targets with mot matching numbers of partitioned are not colocated");
+            throw new ColocationMappingException("Partitioned targets with not matching numbers of partition nodes are not colocated");
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java:200]:
```suggestion
            throw new ColocationMappingException("Partitioned targets with not matching numbers of partitions are not colocated");
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java [hunk 3]
```diff
@@ -188,7 +188,7 @@ static ExecutionTarget colocate(OneOfTarget oneOf, SomeOfTarget someOf) throws C
 
     static ExecutionTarget colocate(PartitionedTarget partitioned, PartitionedTarget otherPartitioned) throws ColocationMappingException {
         if (partitioned.partitionsNodes.length != otherPartitioned.partitionsNodes.length) {
-            throw new ColocationMappingException("Partitioned targets with mot matching numbers of partitioned are not colocated");
+            throw new ColocationMappingException("Partitioned targets with not matching numbers of partition nodes are not colocated");
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java:191]:
```suggestion
            throw new ColocationMappingException("Partitioned targets with not matching numbers of partitions are not colocated");
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java [hunk 4]
```diff
@@ -668,7 +737,7 @@ private static ResultSetMetadata resultSetMetadata(
         );
     }
 
-    private IgniteRel doOptimize(PlanningContext ctx, SqlNode validatedNode, IgnitePlanner planner, CacheKey key) {
+    private IgniteRel doOptimize(PlanningContext ctx, SqlNode validatedNode, IgnitePlanner planner, @Nullable Runnable action) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:740]: let's give new argument more meaningful name, like `onTimeoutAction`, for instance

**Review by korlov42 — APPROVED** [review]
On my laptop results are more noticeable:
```
Before
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert               1    false                25               1  avgt   20  129.715 ± 1.216  us/op

After
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  (replicaCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlInlinedInsert               1    false                25               1  avgt   20  115.613 ± 1.849  us/op
```

Anyways, the patch looks good in general. I left a few minor comments, it would be nice address them before merging.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java (hunk 2 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java (hunk 3 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java]: done

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**